Add wireguard support

shreyagupta30 opened the pull request:
This adds wireguard schema and it's tests. 


Signed-off-by : Shreya Gupta <shreyagupta3011@gmail.com>

shreyagupta30 committed:
schema: Add WireGuard schema

Example:

```yaml
---
interfaces:
  - name: wg0
    type: wireguard
    state: up
    wireguard:
      listen-port: 9000
      private-key: 'exampleZp5x87cNoRJaHdAOzxrxDfDUn7pGmrY/AmzI='
      peers:
        - allowed-ips:
            - 192.0.2.1
          endpoint: 'my-wg.example.com:4001'
          preshared-key: 'qoNbN/6ABe4wWyz4jh+uwX7vqRpNeGEtgAnUbwNjEug='
          public-key: '2Gl0SATbfrrzxfrSkhNoRR9Jg56y533y07KtIVngAk0='
```

Signed-off-by: shreyagupta30 <shreyagupta3011@gmail.com>

ovirt-infra commented:
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

cathay4t commented on libnmstate/schema.py:
```diff
@@ -437,3 +439,14 @@ class Mode:
+    TYPE = InterfaceType.WIREGUARD
+    CONFIG_SUBTREE = "wireguard"
+    FWMARK = "fwmark" 
+    IPV4_AUTO_DEFAULT_ROUTE = "ipv4-auto-default-route"
+    IPV6_AUTO_DEFAULT_ROUTE = "ipv6-auto-default-route"	 
+    LISTEN_PORT = "listen-port" 	 	 
+    MTU = "mtu" 	 	 
```
The MTU is already included as `Interface.MTU`. There is no need to mention it here.

cathay4t commented on libnmstate/schema.py:
```diff
@@ -437,3 +439,14 @@ class Mode:
+    CONFIG_SUBTREE = "wireguard"
+    FWMARK = "fwmark" 
+    IPV4_AUTO_DEFAULT_ROUTE = "ipv4-auto-default-route"
+    IPV6_AUTO_DEFAULT_ROUTE = "ipv6-auto-default-route"	 
```
These two are just a wrapper done by NetworkManager.
The wireguard natively does not have them.

IMHO, user can manually get the same result using `fwmark` and `route rule` as mentioned in https://www.wireguard.com/netns/#routing-all-your-traffic

Please list the use cases which these options are required mandatory.

cathay4t commented on libnmstate/schema.py:
```diff
@@ -437,3 +439,14 @@ class Mode:
+    IPV6_AUTO_DEFAULT_ROUTE = "ipv6-auto-default-route"	 
+    LISTEN_PORT = "listen-port" 	 	 
+    MTU = "mtu" 	 	 
+    PEER_ROUTES = "peer-routes" 	 	 
```
The peer-routes can be done by existing route support of nmstate. No need to depend NM design for this.

cathay4t reviewed (changes requested):
We are not child project of NetworkManager, please don't copy whatever NetworkManager provides to our schema. Nmstate has plugin design which allows other backends.

And nmstate schema should not depend on implementation of certain backend, so you should look on wireguard options itself instead of what NM provides.

The `ipv4-auto-default-route`, `peer-routes` and etc are just convenient stuffs created by NM which can be done via existing nmstate route and route rule schema.

You also missed a lot options:
 * Peer
 * KeepAlive
 * Allowed IP

We will not merge schema-only PR without implementation/test code showing it actual works.

We seems don't have intention to support wireguard server, please also mention that in commit message indication this is only for wireguard client.

ffmancera commented:
> We are not child project of NetworkManager, please don't copy whatever NetworkManager provides to our schema. Nmstate has plugin design which allows other backends.
> 
> And nmstate schema should not depend on implementation of certain backend, so you should look on wireguard options itself instead of what NM provides.
> 

While I do agree on this, I see an issue here. NetworkManager provides certain options and some of them are wrappers that does not exist on wireguard. I will look at them and help @shreyagupta30 on this :-)

> The `ipv4-auto-default-route`, `peer-routes` and etc are just convenient stuffs created by NM which can be done via existing nmstate route and route rule schema.
> 
> You also missed a lot options:
> 
>     * Peer
> 
>     * KeepAlive
> 
>     * Allowed IP
> 

We are going to collect them and add these.

> 
> We will not merge schema-only PR without implementation/test code showing it actual works.

We are not planning on merging them now. This a PR for initial feedback which is nice.

> 
> We seems don't have intention to support wireguard server, please also mention that in commit message indication this is only for wireguard client.

Yep. That is missing.

By the way, if we want to support other options than the supported from NetworkManager we are going to need WIreguard support on Nispor and therefore, generic netlink support on rtnetlink. I will investigate this.

Thank you! @cathay4t @shreyagupta30

ffmancera commented:
Not stale :-)

shreyagupta30 committed:
Add WireGuard support implementation

This patch is adding full support of WireGuard interfaces using
NetworkManager as a backend. Examples:

```yaml
---
interfaces:
  - name: wg0
    type: wireguard
    state: up
    wireguard:
      fwmark: 0
      listen-port: 9000
      private-key: 'exampleZp5x87cNoRJaHdAOzxrxDfDUn7pGmrY/AmzI='
```

```yaml
---
interfaces:
  - name: wg0
    type: wireguard
    state: up
    wireguard:
      listen-port: 9000
      private-key: 'exampleZp5x87cNoRJaHdAOzxrxDfDUn7pGmrY/AmzI='
      peers:
        - allowed-ips:
            - 192.0.2.1
          endpoint: 'my-wg.example.com:4001'
          preshared-key: 'qoNbN/6ABe4wWyz4jh+uwX7vqRpNeGEtgAnUbwNjEug='
          public-key: '2Gl0SATbfrrzxfrSkhNoRR9Jg56y533y07KtIVngAk0='
```

Integration tests addedd.

Signed-off-by: shreyagupta30 <shreyagupta3011@gmail.com>

thom311 commented on libnmstate/ifaces/ifaces.py:
```diff
@@ -44,6 +44,7 @@
 from .vlan import VlanIface
 from .vxlan import VxlanIface
 from .vrf import VrfIface
+from .wireguard import WireguardIface
```
the WireGuard maintainers have a strong opinion about the spelling and styling of WireGuard and the API.

In particular, they wish that `WireGuard` is spelled with upper `G`. NetworkManager's API does that too, at places where there is a CamelCase identifier (but it violates that are places like `NM_SETTING_WIREGUARD_SETTING_NAME` or `nm_wireguard_peer_new()`. By doing that, NetworkManager already violates it's own style, because from `NMSettingWireGuard` one would normally expect `NM_SETTING_WIRE_GUARD_SETTING_NAME`.

They probably would also like to see that the configuration options are spelled like why are in `man wg`. That is, `AllowedIPs` and `PresharedKey` instead of `allowed-ips` and `preshared-key`. However, that style already conflicted with NetworkManager's API, so NetworkManager does not do that either. So never mind that.

It's only about `WireGuard` instead of `Wireguard`.

shreyagupta30 replied:
Hi @thom311, I will make sure to correct it in my next commit. Thanks a lot for the review. :)

ffmancera replied:
@cathay4t What do you think? Should we break our existing style? I do not like the idea.. but it's true that they have strong opinions on that.

cathay4t replied:
For internal stuff like class name, you can follow the guidance of WireGuard. But for public property, please use lower case with dash for consistency

ffmancera committed:
nm.secrets: move get_secrets() to a common file

get_secrets() will be used from WireGuard support. Therefore, it should
be shared in order to do not duplicate code.

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

shreyagupta30 commented:
Not stale :)

ffmancera commented:
Not stale :.-)

ffmancera commented:
Not stale again :(

shreyagupta30 commented:
Not stale

cathay4t commented on libnmstate/nm/ieee_802_1x.py:
```diff
@@ -17,11 +17,11 @@
 
 from .common import NM
 from .common import GLib
+from .utils import get_secrets
```
If we have file called `utils.py`, it will eventually grow bloated. Please use `secret.py` instead.

cathay4t commented on libnmstate/nm/wireguard.py:
```diff
@@ -0,0 +1,145 @@
+def get_info(context, nm_ac):
+    if not nm_ac:
+        return {}
+    nm_profile = nm_ac.get_connection()
```
Even this is for temporary use(we will move to Nispor for querying the kernel interface), but please use `applied_config` if possible. Please check `nm/ipv4.py` for example.

ffmancera replied:
Hi Gris, this won't work. applied_config is a `NM.SimpleConnection` and that class does not contain `get_secrets_async()` that is needed to get the secrets of Wireguard. The class required is `NM.RemoteConnection` which is the one that `nm_ac.get_connection()` is provided. I've been working with Shreya on this but I think unfortunately we cannot use `applied_config` here.

ffmancera replied:
What do you think about my comment @cathay4t?

cathay4t replied:
If applied connection does not provide such info, I am OK with saved configure.

cathay4t commented on libnmstate/nm/connection.py:
```diff
@@ -142,7 +143,6 @@ def create_new_nm_simple_conn(iface, nm_profile):
 
         if wired_setting:
             settings.append(wired_setting)
-
```
This is unrelated change.

cathay4t commented on libnmstate/nm/utils.py:
```diff
@@ -0,0 +1,48 @@
+    try:
+        nm_secrets = nm_profile.get_secrets_finish(result)
+    except GLib.Error as e:
+        context.fail(NmstateLibnmError(f"{action} failed: error={e}"))
```
A `return` is required here.

cathay4t commented on libnmstate/schema.py:
```diff
@@ -459,3 +461,19 @@ class Ieee8021X:
+    FWMARK = "fwmark"
+    LISTEN_PORT = "listen-port"
+    PRIVATE_KEY = "private-key"
+    PEERS = "peers"
```
The `Wireguard.PEERS` should changed to `Wireguard.Peer.CONFIG_SUBTREE` for consistency.

cathay4t commented on libnmstate/schemas/operational-state.yaml:
```diff
@@ -716,6 +717,45 @@ definitions:
+                  public-key:
+                    type: string
+                  flags:
+                    type: integer
```
Both `flags` and `peer.flags` are not used. Please remove them.

cathay4t commented on libnmstate/schemas/operational-state.yaml:
```diff
@@ -716,6 +717,45 @@ definitions:
+          type: object
+          properties:
+            fwmark:
+              type: integer
```
The wireguard wg CLI tool supports string like `0xf1`, please support it also.

cathay4t commented on tests/integration/wireguard_test.py:
```diff
@@ -0,0 +1,139 @@
+        wg_state[Wireguard.LISTEN_PORT] = 9500
+
+        libnmstate.apply(desired_state)
+        assertlib.assert_state(desired_state)
```
Please use `assert_state_match`, in the future there might be more properties added to wireguard, the `assert_state` will fail then.

cathay4t commented on tests/integration/wireguard_test.py:
```diff
@@ -0,0 +1,139 @@
+            },
+        ]
+        libnmstate.apply(desired_state)
+        assertlib.assert_state_match(desired_state)
```
Please use `assert_state_match` instead.

cathay4t commented on tests/integration/wireguard_test.py:
```diff
@@ -0,0 +1,139 @@
+        assertlib.assert_state_match(desired_state)
+
+
+def test_remove_wireguard_peer():
```
Currently, you are trusting `libnmstate.apply()` to do the proper verification. But your patch is ask NM to write a profile, then we verify it by reading this NM profile in verification stage. That does not sounds like trustable.

Could you consider to use `wg showconf` to actual check whether we are doing correct?

cathay4t commented on libnmstate/ifaces/wireguard.py:
```diff
@@ -0,0 +1,48 @@
+from .base_iface import BaseIface
+
+
+class WireguardIface(BaseIface):
```
Need unit test case for this class.

cathay4t reviewed (changes requested):
In summery, these changes required:
 * Rename `utils.py` to `secret.py` to be more specific.
 * Support hex string for `fwmark`.
 * Improve the test to use `wg showconf` for validation.
 * Need unit test case for `WireguardIface`.

shreyagupta30 commented:
not stale

shreyagupta30 commented:
Not stale

ffmancera commented:
This was not stale, we are looking at it. Sorry for the inconvenience.
